osquery_rules/linux/osquery_linux_logins_non_office.py

```diff
@@ -15,11 +15,19 @@ def rule(event):
     host_ip = event['columns'].get('host')
     if not host_ip:
         return False
+    host_ipaddr = ipaddress.IPv4Address(host_ip)
 
-    if ipaddress.IPv4Address(host_ip) not in OFFICE_NETWORK.hosts():
-        return True
+    non_office_logins = []
+    for office_network in OFFICE_NETWORKS:
+        # Check the case where the OFFICE_NETWORK is a /32
+        if office_network.num_addresses == 1:
+            non_office_logins.append(
+                office_network.network_address != host_ipaddr)
+        # Check all possible hosts in the office subnet
+        else:
+            non_office_logins.append(host_ipaddr not in office_network.hosts())
```

Comment:
If you just check for `host_ipaddr not in office_network` on as opposed to `host_ipaddr not in office_network.hosts()` on line 28, you don't need the special case for a /32:

```
>>> import ipaddress
>>> office_networks = [ipaddress.ip_network('192.168.1.100/32'), ipaddress.ip_network('192.168.1.0/24')]
>>> host_ipaddr = ipaddress.IPv4Address('192.168.1.100')
>>> for office_network in office_networks:
...     print(host_ipaddr in office_network)
...
True
True
>>>
```

I'm not sure which one if more memory/time efficient under the hood for large subnets, I imagine they are about similar though. But I think without the special case the code is more clear.

Reply:
Also holds true for public IP address spaces:

```
>>> office_networks = [ipaddress.ip_network('100.100.100.100/32'), ipaddress.ip_network('100.100.100.0/24')]
>>> host_ipaddr = ipaddress.IPv4Address('100.100.100.100')
>>> for office_network in office_networks:
...     print(host_ipaddr in office_network)
...
True
True
>>>
```